proto/prostgen/rules.bzl

```diff
@@ -0,0 +1,325 @@
+    output_dir = ctx.actions.declare_directory(ctx.label.name)
+    lib_rs = ctx.actions.declare_file("%s/lib.rs" % ctx.label.name)
+    args = ctx.actions.args()
+
+    direct_sources = depset(transitive = [
+        depset(f[ProtoInfo].direct_sources)
+        for f in ctx.attr.deps
+    ])
+
+    transitive_sources = depset(transitive = [
+        f[ProstGenInfo].transitive_sources
+        for f in ctx.attr.extern
+    ] + [
+        depset(f[ProtoInfo].transitive_sources)
+        for f in ctx.attr.deps
+    ])
+
+    transitive_externs = depset(transitive = [depset(ctx.attr.extern)] + [
+        f[ProstGenInfo].transitive_externs
+        for f in ctx.attr.extern
+    ])
+
+    extern_descriptors = depset([
+        f[ProstGenInfo].descriptor_set
+        for f in transitive_externs.to_list()
+    ])
+
+    transitive_proto_path = depset(
+        transitive = [
+            f[ProstGenInfo].transitive_proto_path
+            for f in ctx.attr.extern
+        ] + [
+            f[ProtoInfo].transitive_proto_path
+            for f in ctx.attr.deps
+        ],
+    )
+
+    args.add_all(direct_sources)
+    args.add("--output=" + output_dir.path)
+    if ctx.attr.grpc:
+        args.add("--grpc")
+    args.add_all(
+        transitive_externs,
+        map_each = _external_flag,
+    )
+    args.add_all(transitive_proto_path, format_each = "-I%s")
+
+    rustfmt = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"].rustfmt
+    ctx.actions.run(
+        inputs = depset(
+            transitive = [transitive_sources, extern_descriptors],
+        ),
+        env = {
+            # We've patched rustfmt to expect that it can resolve the path to
+            # protoc at runtime via this environment variable, rather than
+            # expecting that this path was available at build time.
+            "PROTOC": ctx.executable._protoc.path,
+            # Since tonic_build has logic to run rustfmt, and does so by default
+            # for that matter, inject rustfmt into PATH, so that
+            # std::process::Command can find it.  This seemed to be more clear
+            # than setting up a separate bazel run action, and removes another
+            # generated file.  A run action can't have the same input and
+            # output file.  This is also compatible with prostgen currently
+            # creating a lib.rs in the output directory, rather than taking in
+            # an output directory and an output file.  The idea with only taking
+            # in an output directory was to try and leave open the possibility
+            # of using include! macros pointing towards the generated files.
+            # This might make things a little easier to read, rather than being
+            # extremely nested by package name.
+            #
+            # May not be perfect either, as the files are formatted before they
+            # get indented to match their module nesting.
+            "PATH": rustfmt.dirname,
```

Comment:
relevant section for tonic-build PR

Reply:
How does https://github.com/hyperium/tonic/pull/571/files solve for this? What's the code change here?

Reply:
Pinging this question again, can you explain what would change should that PR get merged?

Reply:
Could you move to rustfmt as a library to solve this issue?

Reply:
There was a change merged that allows you to specify `RUSTFMT` as a path to a `rustfmt` binary. So this could become: `"RUSTFMT": rustfmt.path`

Reply:
Once 4.0.1 or higher comes out

Reply:
It is out now.

Reply:
cc @purkhusid

Reply:
Is there any progress? It is blocking my project.